add a new awx-manage command `custom_venvs`

rebeccahhh opened the pull request:
add an awx-manage command that gets pip freeze data from custom_venv and outputs to command line stdout

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
part of #7062  - this command is a glorified pip freeze + some extra stuff, people could navigate to each of their custom virtual environments themselves and run a pip freeze, but this allows them to not, and everyone likes their life to be easier. The extra stuff allows users to see the connections that their existing virtual envs have in awx to things like organizations, jobs, inventory updates, and projects.
<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API


##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 19.1.0
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->
This is built off of existing code and there is a line that gets custom venv paths from the settings module, that line does not seem to be working. I have written around that but want to make a note of it.

rebeccahhh committed:
add an awx-manage command that gets pip freeze data from custom_venvs and outputs to command line stdout

remove analytics tests for counts of custom venvs, bump collector version, and remove list of custom venvs from API

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -295,6 +296,7 @@ def get(self, request, format=None):
         data['teams'] = {'url': reverse('api:team_list', request=request), 'total': team_list.count()}
         data['credentials'] = {'url': reverse('api:credential_list', request=request), 'total': credential_list.count()}
         data['job_templates'] = {'url': reverse('api:job_template_list', request=request), 'total': job_template_list.count()}
+        data['custom_venvs'] = {'url': reverse('api:custom_venv_list', request=request)}
```
should this still be here?

rebeccahhh replied:
oh no it should not, thanks

AlanCoding commented on awx/main/management/commands/custom_venvs.py:
```diff
@@ -0,0 +1,30 @@
+        else:
+            venvs = get_custom_venv_choices()
+            for venv in venvs:
+                print(venv)
```
how does the user know what line corresponds to what custom venv?

rebeccahhh replied:
only one venv path can be passed at a time

rebeccahhh replied:
ah, sorry I misunderstood, the path is the only info shared here, I don't know that there's a good way to demonstrate any more than that which path corresponds to which venv identity

AlanCoding replied:
oh right, yeah that flew past me.

So this usage doesn't show any pip freeze data.

rebeccahhh replied:
if you pass nothing/no path in the command it spits out a list of paths for you to view, if you pass one of those paths in it will spit out the pip freeze data

jbradberry commented on awx/main/utils/common.py:
```diff
@@ -887,30 +888,38 @@ def get_current_apps():
+def get_custom_venv_pip_freeze(venv_path):
+    try:
+        freeze_data = subprocess.run([f"{venv_path}/pip", "freeze"], capture_output=True)
+        pip_data = (freeze_data.stdout).decode('ascii')
```
I'm a bit uncomfortable with assuming they are guaranteed to be ASCII.

jbradberry replied:
I tried looking around to see if the output of `pip freeze` would ever contain non-ASCII characters but don't see anything.  The closest I see is that pip supports a UTF-8 encoding for requirements files: https://pip.pypa.io/en/stable/cli/pip_install/#requirements-file-format

rebeccahhh replied:
subprocess outputs to byte format by default, so I have to decode. I picked ascii arbitrarily so I'm not particularly tied to it, that being said, I _highly_ doubt non-ascii characters would ever be present. 
I could decode to utf-8 if we think it's ever going to be problematic

jbradberry replied:
yeah, hard to say.  I would mostly agree, but haven't come up with any hard reason why a package couldn't be named something with some non-ascii characters.

rebeccahhh replied:
will there ever be any `'`? 😂 cause then we've got trouble.

rebeccahhh replied:
swapped to utf-8

AlanCoding commented on awx/main/utils/common.py:
```diff
@@ -887,30 +888,38 @@ def get_current_apps():
+
+    all_venv_paths.append(settings.BASE_VENV_PATH)  # get paths from settings
+
+    for root, dir, files in os.walk('../awx_devel/var/lib/awx/venv/'):  # get paths on machine
```
what's going on with `/awx_devel` here? Looks like something leftover from debugging in the development environment.

rebeccahhh replied:
yep! It's gone now :)

AlanCoding commented on awx/main/utils/common.py:
```diff
@@ -887,30 +888,38 @@ def get_current_apps():
+def get_custom_venv_choices():
     from django.conf import settings
 
-    custom_paths = custom_paths or settings.CUSTOM_VENV_PATHS
```
I don't see how we get around the need for `CUSTOM_VENV_PATHS` which is removed here. Without that, I don't see how this command could make sense.

rebeccahhh replied:
that command actually doesn't return anything anyways 🤷🏻

rebeccahhh replied:
oh my gosh @AlanCoding I actually take that back. I'm so glad you said this because I swapped this call and the one on line 896 which is why that command doesn't return anything. I'll update 🎊

AlanCoding commented on awx/main/management/commands/custom_venvs.py:
```diff
@@ -0,0 +1,30 @@
+            dest='path',
+            type=str,
+            default='',
+            help='run without arguments to see a list of paths, run with one of those paths as an argument and see the pip freeze data',
```
As a matter of wording:

> run without arguments to see a list of paths

I think that should be left for the general command help text to establish (as opposed to the help text about the option).

Will this take either an absolute path or a relative path?

rebeccahhh replied:
it's an absolute path, I'll work on some wording, maybe with @tvo318 so it matches with docs and is helpful to users

shanemcd commented on awx/main/management/commands/custom_venvs.py:
```diff
@@ -0,0 +1,28 @@
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '--path',
```
Since we only have one argument, what do you think about just making it positional?

rebeccahhh replied:
I feel great about that

rebeccahhh replied:
I feel less great now, the way that parser works requires the flags for the variable to be optional, so it would have to be separated into 2 different commands if I want it to be just positional

rebeccahhh replied:
for future reference it is possible, I was just doing it wrong :)

rebeccahhh commented:
just bumped the collector version, logic remains the same

shanemcd committed:
Ignore awx venv

- General usability / readibility improvements

shanemcd committed:
Rename custom venv export command and add usability tweaks

- Uses a positional argument instead of a named arg
- More helpful output

rebeccahhh committed:
split the one command into two for clarity and remove unused imports

rebeccahhh commented:
split this into 2 commands instead of one for clarity `export_custom_venv` and `list_custom_venvs`

shanemcd commented on awx/main/management/commands/export_custom_venv.py:
```diff
@@ -0,0 +1,34 @@
+        parser.add_argument(
+            'path',
+            type=str,
+            nargs='?',
```
I think we should drop this now.

rebeccahhh replied:
the whole thing or just `nargs`?

AlanCoding replied:
I think the `nargs`?

kdelee commented:
~Whatever happened with being able to figure out what venv each JT/org/etc used in the past, so the user could have a clue to what new replacement EE they should use?~
~Is that just staying on the JT/org/etc model but not showing in the UI? can't rememeber~

We discussed and will be adding a 3rd command to list where a venv is used

rebeccahhh committed:
fixed what would be a rather embarrassing misspeeeling

kdelee commented on awx/main/management/commands/export_custom_venv.py:
```diff
@@ -0,0 +1,32 @@
+            if pip_data:
+                print(pip_data)
+                msg = [
+                    'To list all available custom virtual environments run:',
```
Probably should rephrase this...The virtual environments are present, but not "available" for use.

Maybe:
`'To list all (now deprecated) custom virtual environments run:'`

kdelee commented on awx/main/management/commands/list_custom_venvs.py:
```diff
@@ -0,0 +1,33 @@
+
+
+class Command(BaseCommand):
+    """Returns either a list of custom venv paths from the path passed in the argument"""
```
remove `either` has it doesn't make sense in the sentence

kdelee commented:
@rebeccahhh the extra text is helpful to the first time user, but will cause problems to people who might want to use these commands in a small script or redirect the output

For example, if we could just get the raw output without the help text by passing `-q`, I could use bash to loop over the venvs and cat the pip freeze to a file

```
$ for path in $(awx-manage list_custom_venvs -q); do awx-manage export_custom_venv -q $path > $path/freeze.txt; done
```

rebeccahhh committed:
adding in a 3rd command that shows the associated templates, orgs, and invs the venv is tied to

rebeccahhh commented:
moving this back to in progress since we've decided to make this tool a little more robust with a third command

shanemcd commented on awx/main/management/commands/get_custom_venv_associations.py:
```diff
@@ -0,0 +1,35 @@
+            # sanity check here - is path in list
+            path = options.get('path')
+            orgs = [{"name": org.name, "id": org.id} for org in Organization.objects.filter(custom_virtualenv=path)]
+            ujts = [{"name": org.name, "id": org.id} for org in Organization.objects.filter(custom_virtualenv=path)]
```
This says `ujts` but it looks to be a list of `Organization`s?

rebeccahhh replied:
oh that was me copy and pasting and moving too quickly last night!

shanemcd commented on awx/main/management/commands/get_custom_venv_associations.py:
```diff
@@ -0,0 +1,35 @@
+        parser.add_argument(
+            'path',
+            type=str,
+            nargs='?',
```
It seems like this should be a required argument, I think? The `?` here makes it optional. Not sure that's what you want.

rebeccahhh replied:
I currently handle that in the handle function but it's probably better to handle it here.

shanemcd commented on awx/main/management/commands/get_custom_venv_associations.py:
```diff
@@ -0,0 +1,35 @@
+        # look organiztions and unified job templates (which include JTs, workflows, and Inventory updates)
+        super(Command, self).__init__()
+        results = {}
+        if options.get('path'):
```
You can ditch this conditional once you make the argparse change.

shanemcd commented on awx/main/management/commands/get_custom_venv_associations.py:
```diff
@@ -0,0 +1,35 @@
+            results["Organizations"] = orgs
+            results["Unified Job Templates"] = ujts
+            results["Inventory Sources"] = invsrc
+            print(results)
```
Would it be more readable to print yaml here?

rebeccahhh replied:
100%. This whole thing needs a bit of polish, I think yaml would be the most readable

rebeccahhh committed:
split unified job templates out into jts and projects because workflows
don't have custom virtualenvs & rename file so the command is shorter

update copypasta and add sanity check

kdelee commented on awx/main/management/commands/get_custom_venv_associations.py:
```diff
@@ -0,0 +1,35 @@
+            type=str,
+            nargs=1,
+            default='',
+            help='run this with a path to a virtual environment as an argument to see the pip freeze data',
```
update copy pasta

kdelee commented on awx/main/management/commands/get_custom_venv_associations.py:
```diff
@@ -0,0 +1,35 @@
+            proj = [{"name": proj.name, "id": proj.id} for proj in Project.objects.filter(custom_virtualenv=path)]
+            invsrc = [{"name": inv.name, "id": inv.id} for inv in InventorySource.objects.filter(custom_virtualenv=path)]
+            results["Organizations"] = orgs
+            results["Job Templates"] = ujts
```
undefined. use same variable as defined on line 28...in this case, `jts`

kdelee commented on awx/main/management/commands/get_custom_venv_associations.py:
```diff
@@ -0,0 +1,35 @@
+        results = {}
+        if options.get('path'):
+            # sanity check here - is path in list
+            path = options.get('path')[0]
```
I would add validation that this is in list of custom venvs, otherwise we can get tripped up --especially with trailing slash

```
# works as expected
# awx-manage get_custom_venv_associations /tmp/elijahfoo
{'Organizations': [], 'Job Templates': [{'name': 'Demo Job Template', 'id': 7}], 'Projects': [], 'Inventory Sources': []}

# does not
# awx-manage get_custom_venv_associations /tmp/elijahfoo/
{'Organizations': [], 'Job Templates': [], 'Projects': [], 'Inventory Sources': []}
```

AlanCoding commented:
what does the use case for the newer venv associations command? How do users go from there to EE images?

AlanCoding commented on awx/main/management/commands/custom_venv_associations.py:
```diff
@@ -0,0 +1,37 @@
+                results["Job Templates"] = jts
+                results["Projects"] = proj
+                results["Inventory Sources"] = invsrc
+                print(results)
```
an `else:` here would be helpful.

If the user gives a path (don't they have to in order to run the command?) and it's not a valid path, then it would be more sensible to error rather than do nothing and exit silently.

rebeccahhh replied:
I'll add it

AlanCoding replied:
this may already be covered by the `nargs=1`, I just want to set the expectation that giving no arguments doesn't do nothing and return

AlanCoding commented on awx/main/management/commands/list_custom_venvs.py:
```diff
@@ -0,0 +1,33 @@
+            msg = [
+                '',
+                'To export the contents of a (deprecated) virtual environment, ' 'run the following command while supplying the path as an argument:',
+                'awx-manage export_custom_venv /path/to/venv',
```
Also mention the command to get the uses of that venv, and give the syntax that will work for that.

rebeccahhh committed:
add a -q flag to make scripting easier and general improvements for readability

rebeccahhh commented:
recheck

rebeccahhh commented:
@kdelee we've got a third command now and a -q added as well

rebeccahhh committed:
add the conditionals for incorrect paths and helpful info for if a user does that

and remove unused import

rebeccahhh commented:
recheck

rebeccahhh committed:
restructure so the -q is actually completely quiet

rebeccahhh committed:
remove changes to root.py to keep the custom virtualenvs listed in api/config
